raft: transfer leadership away on stepdown

iskettaneh opened the pull request:
This commit does what
https://github.com/cockroachdb/cockroach/pull/133308 started doing.

Basically, when a leader steps down to a learner due to a config change, we now (with this commit) attempt to transfer leadership to the follower with the largest Match.

This opens the door to removing "force campaigning" on the leaseholder after applying a configuration change which has demoted the leader to a learner. This is good since the force campaigning approach causes some layering violation, and could potentially be risky with leader leases.

Release note: None

References: #129796

iskettaneh committed:
raft: transfer leadership away on stepdown

This commit does what
https://github.com/cockroachdb/cockroach/pull/133308 started doing.

Basically, when a leader steps down to a learner due to a config change,
we now (with this commit) attempt to transfer leadership to the follower
with the largest Match.

This opens the door to removing "force campaigning" on the leaseholder
after applying a configuration change which has demoted the leader to a
learner. This is good since the force campaigning approach causes some
layering violation, and could potentially be risky with leader leases.

Release note: None

References: #129796

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141144)
<!-- Reviewable:end -->

iskettaneh committed:
raft: try to campaignTransfer on conf change

When a follower receives a MsgTimeoutNow, it will try to campaign.
However, if it has an unapplied conf change, it can't campaign. This
commit causes the follower to try to campaign again after the conf
change has been applied.

Release note: None

Fixes: #129796
